Bootstrap4 improvements

ddworak opened the pull request.

Kurtzz commented on bootstrap4/.js/src/main/scala/io/udash/bootstrap/tooltip/TooltipUtils.scala:
```diff
@@ -83,29 +91,29 @@ abstract class TooltipUtils[TooltipType <: Tooltip[_, TooltipType]] {
+    boundary: String | dom.Node = "scrollParent",
+    container: Option[String | dom.Node] = None,
+    content: js.Function1[dom.Node, String] | dom.Node = io.udash.emptyStringNode(),
     delay: Delay = Delay(0 millis, 0 millis),
```
`delay` can be just a number too

Kurtzz commented on bootstrap4/.js/src/main/scala/io/udash/bootstrap/tooltip/TooltipUtils.scala:
```diff
@@ -83,29 +91,29 @@ abstract class TooltipUtils[TooltipType <: Tooltip[_, TooltipType]] {
+    content: js.Function1[dom.Node, String] | dom.Node = io.udash.emptyStringNode(),
     delay: Delay = Delay(0 millis, 0 millis),
     html: Boolean = false,
     offset: String = "0",
```
`offset` can be a number too

Kurtzz commented on bootstrap4/.js/src/main/scala/io/udash/bootstrap/utils/BootstrapStyles.scala:
```diff
@@ -672,3 +597,94 @@ object BootstrapStyles {
+  }
+
+  final class Size(val classMarker: String)(implicit enumCtx: EnumCtx) extends AbstractValueEnum
+  object Size extends AbstractValueEnumCompanion[Size] {
```
`xl` missing
Modal accepts xl size: [Modal](https://getbootstrap.com/docs/4.3/components/modal/#optional-sizes)

ddworak replied:
It doesn't on Boostrap 4.1, migration will be another thing.

Kurtzz commented:
### Alert
- `alert-heading` class missing

### Button
- `btn-group-toggle` classmissing ([here](https://getbootstrap.com/docs/4.3/components/buttons/#checkbox-and-radio-buttons))

- toogle can be achive by adding `data-toggle="button" aria-pressed="false"` instead of manually toggle `active'`attribute ([here](https://getbootstrap.com/docs/4.3/components/buttons/#toggle-states))

### Dropdown

- `dropdown-item-text` class missing
- `btn-group` in top level div of the Component is useless
- whether `<button>` or `<a>` shouldn't depends on `dataToggle` property
`data-toggle` is just a css class which adds an arrow to a button. Consider something like a `ButtonTag` 
- there are only methods with `items` as constructor. What about dropdown with some div with [form](https://getbootstrap.com/docs/4.3/components/dropdowns/#form), period picker or [something](https://getbootstrap.com/docs/4.3/components/dropdowns/#text)? 

### Borders

- `rounded-pill` class missing
- rounded size classes missing (`rounded-sm`, `rounded-lg`)

### Shadows
- all classes missing [here](https://getbootstrap.com/docs/4.3/utilities/shadows/)

### Text
- `text-wrap`, `text-break`, `text-reset` are missing

### Typography
- `display-{}` (1,2,3,4) missing [here](https://getbootstrap.com/docs/4.3/content/typography/#display-headings)

ddworak commented:
Added:
- `alert-heading`
- `btn-group-toggle`
- Shadows
- `display-{}`
- `dropdown-item-text` (also to default item factory)
- `Raw(element` dropdown item

Removed:
- `btn-group` in top level dropdown div

I'm not sure about button toggling with `data-toggle` - I'd rather stick with current approach as it's simple to align with the `active` property.

Unavailable in 4.1:
- `rounded-pill`
- `rounded-sm`, `rounded-lg`
- `text-wrap`
- `text-break`
- `text-reset`
- `-xl` for modals

Kurtzz commented:
> Unavailable in 4.1:
> 
> * `rounded-pill`
> * `rounded-sm`, `rounded-lg`
> * `text-wrap`
> * `text-break`
> * `text-reset`
> * `-xl` for modals

Can we upgrade Bootstrap version to 4.3.0 therefore?

ddworak commented:
@Kurtzz there's some unfinished business with 4.2 https://github.com/UdashFramework/udash-core/pull/252, I guess we'll go all the way to whatever's latest once we get to it

Kurtzz commented on bootstrap4/.js/src/main/scala/io/udash/bootstrap/utils/BootstrapStyles.scala:
```diff
@@ -638,6 +566,11 @@ object BootstrapStyles {
   }
 
   object Typography {
+    def display1 = CssStyleName("display1")
```
Should be `display-1` with a dash etc.
